consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/statemachine/QbftBlockHeightManager.java

```diff
@@ -130,19 +131,51 @@ public void handleBlockTimerExpiry(final ConsensusRoundIdentifier roundIdentifie
 
     logValidatorChanges(qbftRound);
 
+    if (roundIdentifier.equals(qbftRound.getRoundIdentifier())) {
+      buildBlockAndMaybePropose(roundIdentifier, qbftRound);
+    } else {
+      LOG.trace(
+          "Block timer expired for a round ({}) other than current ({})",
+          roundIdentifier,
+          qbftRound.getRoundIdentifier());
+    }
+  }
+
+  private void buildBlockAndMaybePropose(
+      final ConsensusRoundIdentifier roundIdentifier, final QbftRound qbftRound) {
+
     // mining will be checked against round 0 as the current round is initialised to 0 above
     final boolean isProposer =
         finalState.isLocalNodeProposerForRound(qbftRound.getRoundIdentifier());
 
-    if (isProposer) {
-      if (roundIdentifier.equals(qbftRound.getRoundIdentifier())) {
-        final long headerTimeStampSeconds = Math.round(clock.millis() / 1000D);
-        qbftRound.createAndSendProposalMessage(headerTimeStampSeconds);
+    final long headerTimeStampSeconds = Math.round(clock.millis() / 1000D);
+    final Block block = qbftRound.createBlock(headerTimeStampSeconds);
+    final boolean blockHasTransactions = !block.getBody().getTransactions().isEmpty();
+    if (blockHasTransactions) {
+      if (isProposer) {
+        LOG.debug("Block has transactions and this node is a proposer so it will send a proposal");
+        qbftRound.sendProposalMessage(block);
+      } else {
+        LOG.debug(
+            "Block has transactions but this node is not a proposer so it will not send a proposal");
+      }
+    } else {
+      final long emptyBlockPeriodSeconds = finalState.getBlockTimer().getEmptyBlockPeriodSeconds();
+      final long emptyBlockPeriodExpiryTime = parentHeader.getTimestamp() + emptyBlockPeriodSeconds;
+      final long nowInSeconds = finalState.getClock().millis() / 1000;
+      if (nowInSeconds >= emptyBlockPeriodExpiryTime) {
+        if (isProposer) {
+          LOG.debug(
+              "Block has no transactions and this node is a proposer so it will send a proposal");
+          qbftRound.sendProposalMessage(block);
+        } else {
+          LOG.debug(
+              "Block has no transactions but this node is not a proposer so it will not send a proposal");
+        }
       } else {
-        LOG.trace(
-            "Block timer expired for a round ({}) other than current ({})",
-            roundIdentifier,
-            qbftRound.getRoundIdentifier());
+        finalState.getRoundTimer().cancelTimer();
+        finalState.getBlockTimer().startEmptyBlockTimer(roundIdentifier, parentHeader);
```

Comment:
Here you are setting the block timer to emptyBlockPeriodSeconds _instead_ of the blockPeriodSeconds, so despite having transactions in the pool, no block proposal is attempted until emptyBlockPeriodSeconds has passed since the last block, at the earliest.

Reply:
As mentioned above, GoQuorum QBFT fundamentally differs because GoQuorum follows geth's approach of having a separate thread with a tight block building loop. Besu QBFT does not have this block building loop running in the background, instead it is event-driven using timers.

It would be a large change to Besu QBFT to support a tight block building loop, but I think for better liveness, ensuring we do something every block period, even if it's just skipping an empty block is a better approach than waiting the full emptyBlockPeriodSeconds.

Reply:
For example, consider if a user were to configure `blockPeriodSeconds: 1` along with `emptyBlockPeriodSeconds: 600` - I think the user would complain that they had to wait 10 minutes for their transactions to go through

Reply:
Yes, we already reached that conclusion. Please read my last comment.